src/Methods/Pipes/BuilderAfterRelations.php

```diff
@@ -0,0 +1,90 @@
+<?php
+
+declare(strict_types=1);
+
+/**
+ * This file is part of Larastan.
+ *
+ * (c) Nuno Maduro <enunomaduro@gmail.com>
+ *
+ *  For the full copyright and license information, please view the LICENSE
+ *  file that was distributed with this source code.
+ */
+
+namespace NunoMaduro\Larastan\Methods\Pipes;
+
+use Closure;
+use Illuminate\Database\Eloquent\Builder as EloquentBuilder;
+use Illuminate\Database\Eloquent\Model;
+use Illuminate\Database\Eloquent\Relations\Relation;
+use Illuminate\Database\Query\Builder as QueryBuilder;
+use NunoMaduro\Larastan\Contracts\Methods\PassableContract;
+use NunoMaduro\Larastan\Contracts\Methods\Pipes\PipeContract;
+use NunoMaduro\Larastan\Reflection\EloquentBuilderMethodReflection;
+use NunoMaduro\Larastan\Reflection\ModelScopeMethodReflection;
+use PHPStan\Reflection\ParametersAcceptorSelector;
+use PHPStan\Type\ObjectType;
+
+final class BuilderAfterRelations implements PipeContract
+{
+    /**
+     * {@inheritdoc}
+     */
+    public function handle(PassableContract $passable, Closure $next): void
+    {
+        $classReflection = $passable->getClassReflection();
+        $isRelationSubclass = $classReflection->isSubclassOf(Relation::class);
+
+        $found = false;
+
+        if ($isRelationSubclass && ! $classReflection->isAbstract()) {
+            if (($returnMethodReflection = $this->getMethodReflectionForBuilder($passable, EloquentBuilder::class))
+                || ($returnMethodReflection = $this->getMethodReflectionForBuilder($passable, QueryBuilder::class))) {
+                $passable->setMethodReflection($returnMethodReflection);
+                $found = true;
+            } else {
+                $passable->setMethodReflection(
+                    new ModelScopeMethodReflection(
+                        $passable->getMethodName(),
+                        $passable->getBroker()->getClass(Model::class),
+                        $classReflection
+                    )
+                );
+
+                $found = true;
+            }
+        }
+
+        if (! $found) {
+            $next($passable);
+        }
+    }
+
+    /**
+     * @param class-string $builderClass
+     */
+    private function getMethodReflectionForBuilder(PassableContract $passable, string $builderClass): ?EloquentBuilderMethodReflection
+    {
+        $eloquentBuilder = $passable->getBroker()->getClass($builderClass);
+
+        if (! $eloquentBuilder->hasNativeMethod($passable->getMethodName())) {
```

Comment:
You can do an additional check here like `! Str::startsWith($passable->getMethodName(), 'where')` (not sure of the boolean logic. You can check) So we will allow the dynamic wheres also.

Reply:
Yeah that won't work. You will then get the following error:
 `Method Illuminate\\Database\\Eloquent\\Model::whereFoo() invoked with 1 parameter, 0 required.`. 

If you take a look at [`BuilderDynamicWheres`](https://github.com/nunomaduro/larastan/blob/master/src/Methods/Pipes/BuilderDynamicWheres.php) you will see that there is much more needed to support those dynamic where's. 

So maybe we should create a `BuilderHelper` where we can return method reflection for that dynamic where. And call that in both `BuilderDynamicWhere` as `BuilderAfterRelations`. And while writing this I think we should add support for [local macros](https://github.com/nunomaduro/larastan/blob/master/src/Methods/Pipes/BuilderLocalMacros.php) on the relationships. Which we can also add to that `BuilderHelper`

Reply:
Ok, you can give that a go :+1:

Reply:
I think this is a nice solution. `BuilderLocalMacros` is depending on the model which [as discussed](https://github.com/nunomaduro/larastan/issues/409#issuecomment-571961110) I don't have access to.